ucp/2023-12-irregular-types.md

```diff
@@ -0,0 +1,269 @@
+
+### What about ARM?
+
+ARM is the inspiration for UCP's design, and we kept this approach.
+
+```bash
+az rest --method get --url "/subscriptions/84bed6f7-783b-4cb5-a3c5-d42837c57770/resourceGroups/example-group?api-version=2020-06-01"
+```
+
+```json
+{
+  "id": "/subscriptions/84bed6f7-783b-4cb5-a3c5-d42837c57770/resourceGroups/example-group",
+  "name": "example-group",
+  "type": "Microsoft.Resources/resourceGroups",
+  "location": "eastus",
+  "tags": {},
+  "properties": {
+    "provisioningState": "Succeeded"
+  }
+}
+```
+
+The regular form of this ID also results in a 404 on ARM. 
+
+```bash
+az rest --method get --url "/subscriptions/84bed6f7-783b-4cb5-a3c5-d42837c57770/providers/Microsoft.Resources/resourceGroups/example-group?api-version=2020-06-01"
+```
+
+
+### Design details / API design
+
+**This section proposes design principles for these cases.**
+
+#### Avoid irregular APIs
+
+First of all, we should limit the set of cases where we have to use irregular APIs. We can list all of these cases and the are all part of UCP.
+
+- Planes/Tenants: eg. `/planes/radius/local`
+- Resource Groups: eg. `/planes/radius/local/resourceGroups/example-group`
+- Resource Providers (type discovery): eg. `/planes/radius/local/resourceGroups/example-group/providers`
+
+We should forbid resource providers from adding their own irregular APIs. UCP is the front-proxy for resource providers, and we can validate during proxying that incoming requests match a regular pattern.
+
+#### Standardize on regular APIs
+
+For the cases listed above we should provide these as regular APIs in addition to the irregular APIs.
+
+- Planes/Tenants: eg. `/planes/radius/local` -> `/planes/providers/System.Planes/radius/local`
+- Resource Groups: eg. `/planes/radius/local/resourceGroups/example-group` -> `/planes/radius/local/providers/System.Resources/resourceGroups/example-group`
+- Resource Providers (type discovery): eg. `/planes/radius/local/resourceGroups/example-group/providers` -> `/planes/radius/local/System.Resources/providers`
+
+For the irregular types, the URL segment they provide (eg. `resourceGroups`) must match the unqualified resource type. 
+
+Importantly, we should standardize on the **regular** format for resource IDs. For the resource types that represent a **scope** we can also include the scope format in a new property for use in client code.
+
+**Looking for input on the right property name. I'm not sure scope is a great one.**
+
+```jsonc
+{
+  "id": "/planes/radius/local/providers/System.Resource/resourceGroups/default", // Changed
+  "scope": "/planes/radius/local/resourcegroups/default", // Added
```

Comment:
Would we also add the `scope` field to all resources? Such that when you do a GET on a container you get the `id` of the container (no change), and you also get the `scope` as well?

Reply:
This is probably the part of the proposal I am the most unsure about 😆 

It's nice today that you can combine the `id` of a scope (like a resource group) with other parts to make a full `id`.

Maybe this belongs in `properties` instead, and only for things that are scope-like.